[JSC] Refine JSMap and JSSet with CloseTable

hyjorc1 opened the pull request:
#### b451fca37277a0b56042695981ba0efda1795d80
<pre>
[JSC] Refine JSMap and JSSet with CloseTable
<a href="https://bugs.webkit.org/show_bug.cgi?id=274415">https://bugs.webkit.org/show_bug.cgi?id=274415</a>
<a href="https://rdar.apple.com/120931591">rdar://120931591</a>

Reviewed by Yusuke Suzuki.

This patch enhances the implementation of JSMap and JSSet by replacing
the existing storage mechanism (HashMapImpl&lt;HashMapBucket&lt;Data&gt;&gt;) with
an OrderedHashTable. The new OrderedHashTable flattens the CloseTable[1]
structure into a single array, optimizing memory usage and improving
cache locality.

Microbenchmark results with 20 iterations:

                                                without                     with

js-map-set-int                            0.5624+-0.0070     ^      0.4446+-0.0069        ^ definitely 1.2651x faster
js-map-get-string-no-dfg-no-inline        5.3237+-0.0117     ^      5.1858+-0.0176        ^ definitely 1.0266x faster
js-map-get-int                            1.3820+-0.0118     ^      1.2287+-0.0127        ^ definitely 1.1247x faster
js-map-set-int-no-dfg-no-inline           0.6037+-0.0069     ^      0.4741+-0.0073        ^ definitely 1.2734x faster
js-map-get-string                         1.7580+-0.0118     ^      1.6399+-0.0158        ^ definitely 1.0721x faster
js-map-set-string-no-dfg-no-inline        0.8514+-0.0087     ^      0.7264+-0.0080        ^ definitely 1.1722x faster
js-map-get-int-no-dfg-no-inline           3.3643+-0.0109     ^      3.2791+-0.0116        ^ definitely 1.0260x faster
js-map-set-string                         0.8092+-0.0092     ^      0.6811+-0.0091        ^ definitely 1.1881x faster

&lt;geometric&gt;                               1.3356+-0.0070     ^      1.1714+-0.0075        ^ definitely 1.1402x faster

[1] <a href="https://wiki.mozilla.org/User">https://wiki.mozilla.org/User</a>:Jorend/Deterministic_hash_tables

* JSTests/microbenchmarks/js-map-get-int-no-dfg-no-inline.js: Added.
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-int.js: Added.
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-string-no-dfg-no-inline.js: Added.
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-string.js: Added.
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-set-int-no-dfg-no-inline.js: Added.
(testSet):
* JSTests/microbenchmarks/js-map-set-int.js: Added.
(testSet):
* JSTests/microbenchmarks/js-map-set-string-no-dfg-no-inline.js: Added.
(testSet):
* JSTests/microbenchmarks/js-map-set-string.js: Added.
(testSet):
* JSTests/stress/map-iterators-next-2.js: Added.
(assert):
* JSTests/stress/set-iterators-next-2.js: Added.
(assert):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/MapIteratorPrototype.js:
(next):
(linkTimeConstant.mapIteratorNext): Deleted.
* Source/JavaScriptCore/builtins/MapPrototype.js:
(forEach):
* Source/JavaScriptCore/builtins/SetIteratorPrototype.js:
(next):
(linkTimeConstant.setIteratorNext): Deleted.
* Source/JavaScriptCore/builtins/SetPrototype.js:
(forEach):
(isSubsetOf):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
(JSC::BytecodeIntrinsicRegistry::orderedHashTableSentinelValue):
(JSC::BytecodeIntrinsicRegistry::sentinelMapBucketValue): Deleted.
(JSC::BytecodeIntrinsicRegistry::sentinelSetBucketValue): Deleted.
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecode/SpeculatedType.cpp:
(JSC::speculationFromClassInfoInheritance):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::mapIteratorInternalFieldIndex):
(JSC::setIteratorInternalFieldIndex):
* Source/JavaScriptCore/dfg/DFGAbstractHeap.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasHeapPrediction):
(JSC::DFG::Node::hasBucketOwnerType):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::SafeToExecuteEdge::operator()):
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetMapIndexImpl):
(JSC::DFG::SpeculativeJIT::compileMapKeyIndex):
(JSC::DFG::SpeculativeJIT::compileMapValue):
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGUseKind.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGUseKind.h:
(JSC::DFG::typeFilterFor):
(JSC::DFG::isCell):
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h:
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
(Inspector::cloneMapIteratorObject):
(Inspector::cloneSetIteratorObject):
(Inspector::JSInjectedScriptHost::iteratorEntries):
* Source/JavaScriptCore/runtime/HashMapHelper.h: Added.
(JSC::areKeysEqual):
(JSC::normalizeMapKey):
(JSC::wangsInt64Hash):
(JSC::jsMapHash):
(JSC::jsMapHashImpl):
(JSC::jsMapHashForAlreadyHashedValue):
(JSC::concurrentJSMapHash):
(JSC::shouldShrink):
(JSC::shouldRehash):
(JSC::nextCapacity):
* Source/JavaScriptCore/runtime/HashMapImpl.cpp: Removed.
* Source/JavaScriptCore/runtime/HashMapImpl.h: Removed.
* Source/JavaScriptCore/runtime/HashMapImplInlines.h: Removed.
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSCJSValue.h:
(JSC::OrderedHashTableTraits::set):
(JSC::OrderedHashTableTraits::increment):
(JSC::OrderedHashTableTraits::decrement):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::createTuple):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.h:
* Source/JavaScriptCore/runtime/JSMap.h:
* Source/JavaScriptCore/runtime/JSMapInlines.h:
* Source/JavaScriptCore/runtime/JSMapIterator.cpp:
(JSC::JSMapIterator::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSMapIterator::createPair): Deleted.
* Source/JavaScriptCore/runtime/JSMapIterator.h:
* Source/JavaScriptCore/runtime/JSSet.h:
* Source/JavaScriptCore/runtime/JSSetInlines.h:
* Source/JavaScriptCore/runtime/JSSetIterator.cpp:
(JSC::JSSetIterator::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSSetIterator::createPair): Deleted.
* Source/JavaScriptCore/runtime/JSSetIterator.h:
* Source/JavaScriptCore/runtime/MapConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/MapConstructor.h:
* Source/JavaScriptCore/runtime/MapPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::createMapIteratorObject):
* Source/JavaScriptCore/runtime/OrderedHashTable.cpp: Copied from Source/JavaScriptCore/runtime/JSMapInlines.h.
(JSC::OrderedHashTable&lt;Traits&gt;::visitChildrenImpl):
* Source/JavaScriptCore/runtime/OrderedHashTable.h: Added.
(JSC::OrderedHashTable::OrderedHashTable):
(JSC::OrderedHashTable::offsetOfButterfly):
(JSC::OrderedHashTable::finishCreation):
(JSC::OrderedHashTable::getKeyIndex):
(JSC::OrderedHashTable::has):
(JSC::OrderedHashTable::add):
(JSC::OrderedHashTable::addNormalized):
(JSC::OrderedHashTable::remove):
(JSC::OrderedHashTable::removeNormalized):
(JSC::OrderedHashTable::size):
(JSC::OrderedHashTable::clear):
(JSC::OrderedHashTable::materializeIfNeeded):
(JSC::OrderedHashTable::storageOrSentinel):
(JSC::OrderedHashTable::storageRef):
(JSC::OrderedHashMap::OrderedHashMap):
(JSC::OrderedHashMap::getImpl):
(JSC::OrderedHashMap::get):
(JSC::OrderedHashMap::createSentinel):
(JSC::OrderedHashMap::createDeletedValue):
(JSC::OrderedHashSet::OrderedHashSet):
* Source/JavaScriptCore/runtime/OrderedHashTableHelper.h: Added.
(JSC::OrderedHashTableHelper::toNumber):
(JSC::OrderedHashTableHelper::asNumber):
(JSC::OrderedHashTableHelper::toJSValue):
(JSC::OrderedHashTableHelper::isDeleted):
(JSC::OrderedHashTableHelper::isValidTableIndex):
(JSC::OrderedHashTableHelper::invalidTableIndex):
(JSC::OrderedHashTableHelper::get):
(JSC::OrderedHashTableHelper::slot):
(JSC::OrderedHashTableHelper::set):
(JSC::OrderedHashTableHelper::setWithWriteBarrier):
(JSC::OrderedHashTableHelper::aliveEntryCountIndex):
(JSC::OrderedHashTableHelper::deletedEntryCountIndex):
(JSC::OrderedHashTableHelper::capacityIndex):
(JSC::OrderedHashTableHelper::iterationEntryIndex):
(JSC::OrderedHashTableHelper::aliveEntryCount):
(JSC::OrderedHashTableHelper::deletedEntryCount):
(JSC::OrderedHashTableHelper::usedCapacity):
(JSC::OrderedHashTableHelper::capacity):
(JSC::OrderedHashTableHelper::iterationEntry):
(JSC::OrderedHashTableHelper::incrementAliveEntryCount):
(JSC::OrderedHashTableHelper::decrementAliveEntryCount):
(JSC::OrderedHashTableHelper::incrementDeletedEntryCount):
(JSC::OrderedHashTableHelper::bucketCount):
(JSC::OrderedHashTableHelper::hashTableStartIndex):
(JSC::OrderedHashTableHelper::hashTableEndIndex):
(JSC::OrderedHashTableHelper::bucketIndex):
(JSC::OrderedHashTableHelper::dataTableSize):
(JSC::OrderedHashTableHelper::dataTableStartIndex):
(JSC::OrderedHashTableHelper::dataTableEndIndex):
(JSC::OrderedHashTableHelper::entryDataStartIndex):
(JSC::OrderedHashTableHelper::setKeyOrValueData):
(JSC::OrderedHashTableHelper::deleteData):
(JSC::OrderedHashTableHelper::addToChain):
(JSC::OrderedHashTableHelper::tableSize):
(JSC::OrderedHashTableHelper::tableSizeSlow):
(JSC::OrderedHashTableHelper::isObsolete):
(JSC::OrderedHashTableHelper::nextTable):
(JSC::OrderedHashTableHelper::setNextTable):
(JSC::OrderedHashTableHelper::deletedEntriesStartIndex):
(JSC::OrderedHashTableHelper::isCleared):
(JSC::OrderedHashTableHelper::setClearedTableSentinel):
(JSC::OrderedHashTableHelper::tryCreate):
(JSC::OrderedHashTableHelper::copyImpl):
(JSC::OrderedHashTableHelper::copy):
(JSC::OrderedHashTableHelper::rehash):
(JSC::OrderedHashTableHelper::clear):
(JSC::OrderedHashTableHelper::find):
(JSC::OrderedHashTableHelper::expandIfNeeded):
(JSC::OrderedHashTableHelper::addImpl):
(JSC::OrderedHashTableHelper::add):
(JSC::OrderedHashTableHelper::addNormalized):
(JSC::OrderedHashTableHelper::shrinkIfNeed):
(JSC::OrderedHashTableHelper::removeImpl):
(JSC::OrderedHashTableHelper::remove):
(JSC::OrderedHashTableHelper::removeNormalized):
(JSC::OrderedHashTableHelper::transit):
(JSC::OrderedHashTableHelper::transitAndNext):
(JSC::OrderedHashTableHelper::getKey):
(JSC::OrderedHashTableHelper::getValue):
(JSC::OrderedHashTableHelper::nextAndUpdateIterationEntry):
(JSC::OrderedHashTableHelper::getIterationEntry):
(JSC::OrderedHashTableHelper::getIterationEntryKey):
(JSC::OrderedHashTableHelper::getIterationEntryValue):
* Source/JavaScriptCore/runtime/SetConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/SetConstructor.h:
* Source/JavaScriptCore/runtime/SetPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::createSetIteratorObject):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::orderedHashTableDeletedValueSlow):
(JSC::VM::orderedHashTableSentinelSlow):
(JSC::VM::visitAggregateImpl):
(JSC::VM::sentinelSetBucketSlow): Deleted.
(JSC::VM::sentinelMapBucketSlow): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::orderedHashTableDeletedValue):
(JSC::VM::orderedHashTableSentinel):
(JSC::VM::sentinelSetBucket): Deleted.
(JSC::VM::sentinelMapBucket): Deleted.
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
* Source/JavaScriptCore/runtime/WeakMapImplInlines.h:
* Source/JavaScriptCore/runtime/WeakSetPrototype.cpp:
* Source/WebCore/bindings/js/JSDOMMapLike.cpp:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::serialize):

Canonical link: <a href="https://commits.webkit.org/280734@main">https://commits.webkit.org/280734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c807c1a7ca4f49142dd85d17ed5769b4736ac0eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57098 "13 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36426 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9573 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60718 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7541 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59226 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/44434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7731 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46238 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5302 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59128 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34305 "Found 60 new test failures: accessibility/accessibility-node-reparent.html accessibility/ancestor-computation.html accessibility/announcement-notification.html accessibility/aria-braillelabel.html accessibility/aria-brailleroledescription.html accessibility/aria-checked-mixed-value.html accessibility/aria-current-state-changed-notification.html accessibility/aria-current.html accessibility/aria-describedby-on-input.html accessibility/aria-description.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49287 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27099 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31087 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html imported/w3c/web-platform-tests/FileAPI/FileReader/Progress_event_bubbles_cancelable.html imported/w3c/web-platform-tests/FileAPI/FileReader/workers.html imported/w3c/web-platform-tests/FileAPI/FileReaderSync.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-dom.window.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-endings.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6617 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6546 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50190 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53197 "Build was cancelled. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (failure); re-run-api-tests (cancelled)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6887 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62399 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/56340 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1011 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6987 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53496 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1015 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49340 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53588 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12705 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/856 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78100 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32255 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12941 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33340 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34425 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33086 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTableAccessor.h:
```diff
@@ -0,0 +1,825 @@
+// Note that all elements in the JSImmutableButterfly are in JSValue type. However, only the key and value in the DataTable are real JSValues.
+// The others are used as unsigned integers wrapped by JSValue.
+template<typename Traits>
+class OrderedHashTableAccessor : public JSImmutableButterfly {
```
I don't want to inherit `JSImmutableButterfly`. It is very special object, and nobody should inherit it (JSImmutableButterfly is specially handled GC object. So making inheritance causes super complex issues.)
Can you just define all helpers as a static function taking `JSImmutableButterfly&` instead? Inheriting JSImmutableButterfly is a bit too dangerous.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -2862,7 +2862,7 @@ SerializationReturnCode CloneSerializer::serialize(JSValue in)
                     break;
                 write(MapObjectTag);
-                JSMapIterator* iterator = JSMapIterator::create(vm, m_lexicalGlobalObject->mapIteratorStructure(), inMap, IterationKind::Entries);
+                JSMapIterator* iterator = JSMapIterator::create(m_lexicalGlobalObject, m_lexicalGlobalObject->mapIteratorStructure(), inMap, IterationKind::Entries);
```
This can throw an error, right? It should handle errors.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -2908,7 +2908,7 @@ SerializationReturnCode CloneSerializer::serialize(JSValue in)
                     break;
                 write(SetObjectTag);
-                JSSetIterator* iterator = JSSetIterator::create(vm, m_lexicalGlobalObject->setIteratorStructure(), inSet, IterationKind::Keys);
+                JSSetIterator* iterator = JSSetIterator::create(m_lexicalGlobalObject, m_lexicalGlobalObject->setIteratorStructure(), inSet, IterationKind::Keys);
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -2466,6 +2468,232 @@ void SpeculativeJIT::emitBranch(Node* node)
+    JSValueRegsTemporary result(this);
+
+    GPRReg butterflyStorageGPR = butterflyStorage.gpr();
+    JSValueRegs resultRegs = result.regs();
```
Let's first list all Operands / Temporaries, and calling gpr() etc. later and then call speculate etc.
```
SpeculateCellOperand map(this, node->child1());
SpeculateInt32Operand keyIndex(this, node->child2());
GPRTemporary butterflyStorage(this);
JSValueRegsTemporary result(this);

GPRReg mapGPR = map.gpr();
GPRReg keyIndexGPR = keyIndex.gpr();
GPRReg butterflyStorageGPR = butterflyStorage.gpr();
JSValueRegs resultRegs = result.regs();

speculateMapObject(node->child1(), mapGPR);
```

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -2466,6 +2468,232 @@ void SpeculativeJIT::emitBranch(Node* node)
+    GPRTemporary result(this);
+
+    JSValueRegsTemporary entryKeyIndex(this);
+    JSValueRegsTemporary entryKey(this);
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -11951,111 +11951,21 @@ void SpeculativeJIT::compileNewSymbol(Node* node)
-    addSlowPathGenerator(slowPathCall(slowPath, this, operationNewMap, resultGPR, TrustedImmPtr(&vm()), node->structure()));
-
+    callOperation(operationNewMap, resultGPR, TrustedImmPtr(&vm()), node->structure());
+    exceptionCheck();
```
Can we make this inline?

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -11951,111 +11951,21 @@ void SpeculativeJIT::compileNewSymbol(Node* node)
-    addSlowPathGenerator(slowPathCall(slowPath, this, operationNewSet, resultGPR, TrustedImmPtr(&vm()), node->structure()));
-
+    callOperation(operationNewSet, resultGPR, TrustedImmPtr(&vm()), node->structure());
+    exceptionCheck();
```
Ditto.

hyjorc1 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -16592,42 +16650,31 @@ IGNORE_CLANG_WARNINGS_END
 
-        LBasicBlock slowCase = m_out.newBlock();
-        LBasicBlock continuation = m_out.newBlock();
+        // LBasicBlock slowCase = m_out.newBlock();
```
I need to remove this.

hyjorc1 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -16558,32 +16629,19 @@ IGNORE_CLANG_WARNINGS_END
 
-        LBasicBlock slowCase = m_out.newBlock();
-        LBasicBlock continuation = m_out.newBlock();
+        // LBasicBlock slowCase = m_out.newBlock();
```
I need to remove this.

hyjorc1 commented on Source/JavaScriptCore/runtime/OrderedHashTableAccessor.h:
```diff
@@ -0,0 +1,825 @@
+namespace JSC {
+
+struct MapTraits {
+    // DataTable:  [ Key_0, Val_0, NextInChain_0, Key_1, Val_1, NextKeyIndexInChain_1, ... ]
```
I need to fix this.

hyjorc1 commented on Source/JavaScriptCore/runtime/OrderedHashTableAccessor.h:
```diff
@@ -0,0 +1,825 @@
+};
+
+struct SetTraits {
+    // DataTable:  [ Key_0, NextInChain_0, Key_1, NextKeyIndexInChain_1, ... ]
```
I need to fix this.

Constellation commented on Source/JavaScriptCore/runtime/MapPrototype.cpp:
```diff
@@ -124,7 +124,8 @@ JSC_DEFINE_HOST_FUNCTION(mapProtoFuncClear, (JSGlobalObject* globalObject, CallF
-    map->clear(vm);
+    map->clear(globalObject);
+    RETURN_IF_EXCEPTION(scope, JSValue::encode(jsUndefined()));
     return JSValue::encode(jsUndefined());
```
Let's do,

```
scope.release();
map->clear(globalObject);
return JSValue::encode(jsUndefined());
```

Constellation commented on Source/JavaScriptCore/runtime/SetPrototype.cpp:
```diff
@@ -141,7 +140,8 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncClear, (JSGlobalObject* globalObject, CallF
-    set->clear(vm);
+    set->clear(globalObject);
+    RETURN_IF_EXCEPTION(scope, JSValue::encode(jsUndefined()));
     return JSValue::encode(jsUndefined());
```
Let's do,

```
scope.release();
set->clear(globalObject);
return JSValue::encode(jsUndefined());
```

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTable.h:
```diff
@@ -0,0 +1,233 @@
+    ALWAYS_INLINE JSValue get(JSGlobalObject* globalObject, JSValue key, uint32_t hash)
+    {
+        VM& vm = getVM(globalObject);
+        auto scope = DECLARE_THROW_SCOPE(vm);
```
Let's remove this scope.

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTable.h:
```diff
@@ -0,0 +1,233 @@
+        JSValue result = getImpl(globalObject, [&](Storage& storage) ALWAYS_INLINE_LAMBDA {
+            return Helper::find(globalObject, storage, key, hash);
+        });
+        RETURN_IF_EXCEPTION(scope, { });
```
Let's remove this. Since this function's only error-throwing one is getImpl.

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTable.h:
```diff
@@ -0,0 +1,233 @@
+    ALWAYS_INLINE JSValue get(JSGlobalObject* globalObject, JSValue key)
+    {
+        VM& vm = getVM(globalObject);
+        auto scope = DECLARE_THROW_SCOPE(vm);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTable.h:
```diff
@@ -0,0 +1,233 @@
+        JSValue result = getImpl(globalObject, [&](Storage& storage) ALWAYS_INLINE_LAMBDA {
+            return Helper::find(globalObject, storage, key);
+        });
+        RETURN_IF_EXCEPTION(scope, { });
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTable.h:
```diff
@@ -0,0 +1,233 @@
+    ALWAYS_INLINE bool remove(JSGlobalObject* globalObject, JSValue key)
+    {
+        VM& vm = getVM(globalObject);
+        auto scope = DECLARE_THROW_SCOPE(vm);
```
You do not need this.

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTable.h:
```diff
@@ -0,0 +1,233 @@
+        auto scope = DECLARE_THROW_SCOPE(vm);
+
+        if (m_storage)
+            RELEASE_AND_RETURN(scope, Helper::remove(globalObject, this, storageRef(), key));
```
Let's just return `Helper::remove(globalObject, this, storageRef(), key)`.

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTable.h:
```diff
@@ -0,0 +1,233 @@
+        auto scope = DECLARE_THROW_SCOPE(vm);
+
+        if (m_storage)
+            RELEASE_AND_RETURN(scope, Helper::removeNormalized(globalObject, this, storageRef(), key, hash));
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTable.h:
```diff
@@ -0,0 +1,233 @@
+    {
+        VM& vm = getVM(globalObject);
+        auto scope = DECLARE_THROW_SCOPE(vm);
+
```
Remove this.

Constellation commented on Source/JavaScriptCore/runtime/OrderedHashTable.h:
```diff
@@ -0,0 +1,233 @@
+
+        if (m_storage) {
+            auto result = Helper::find(globalObject, storageRef(), key, hash);
+            RELEASE_AND_RETURN(scope, result.entryKeyIndex);
```
Let's just return `return Helper::find(globalObject, storageRef(), key, hash).entryKeyIndex;`

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
```diff
@@ -3824,28 +3825,66 @@ void SpeculativeJIT::compile(Node* node)
+        auto operation = isMapObjectUse ? operationMapKeyIndex : operationSetKeyIndex;
+        callOperation(operation, resultGPR, LinkableConstant::globalObject(*this, node), mapGPR, keyRegs, hashGPR);
+        exceptionCheck();
+        strictInt32Result(resultGPR, node);
```
Let's change it to new Operation form which does exception check with returned value.

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[JSC] Refine JSMap and JSSet with CloseTable
https://bugs.webkit.org/show_bug.cgi?id=274415
rdar://120931591

Reviewed by Yusuke Suzuki.

This patch enhances the implementation of JSMap and JSSet by replacing
the existing storage mechanism (HashMapImpl<HashMapBucket<Data>>) with
an OrderedHashTable. The new OrderedHashTable flattens the CloseTable[1]
structure into a single array, optimizing memory usage and improving
cache locality.

Microbenchmark results with 20 iterations:

                                                without                     with

js-map-set-int                            0.5624+-0.0070     ^      0.4446+-0.0069        ^ definitely 1.2651x faster
js-map-get-string-no-dfg-no-inline        5.3237+-0.0117     ^      5.1858+-0.0176        ^ definitely 1.0266x faster
js-map-get-int                            1.3820+-0.0118     ^      1.2287+-0.0127        ^ definitely 1.1247x faster
js-map-set-int-no-dfg-no-inline           0.6037+-0.0069     ^      0.4741+-0.0073        ^ definitely 1.2734x faster
js-map-get-string                         1.7580+-0.0118     ^      1.6399+-0.0158        ^ definitely 1.0721x faster
js-map-set-string-no-dfg-no-inline        0.8514+-0.0087     ^      0.7264+-0.0080        ^ definitely 1.1722x faster
js-map-get-int-no-dfg-no-inline           3.3643+-0.0109     ^      3.2791+-0.0116        ^ definitely 1.0260x faster
js-map-set-string                         0.8092+-0.0092     ^      0.6811+-0.0091        ^ definitely 1.1881x faster

<geometric>                               1.3356+-0.0070     ^      1.1714+-0.0075        ^ definitely 1.1402x faster

[1] https://wiki.mozilla.org/User:Jorend/Deterministic_hash_tables

* JSTests/microbenchmarks/js-map-get-int-no-dfg-no-inline.js: Added.
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-int.js: Added.
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-string-no-dfg-no-inline.js: Added.
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-get-string.js: Added.
(testSet):
(testGet):
* JSTests/microbenchmarks/js-map-set-int-no-dfg-no-inline.js: Added.
(testSet):
* JSTests/microbenchmarks/js-map-set-int.js: Added.
(testSet):
* JSTests/microbenchmarks/js-map-set-string-no-dfg-no-inline.js: Added.
(testSet):
* JSTests/microbenchmarks/js-map-set-string.js: Added.
(testSet):
* JSTests/stress/map-iterators-next-2.js: Added.
(assert):
* JSTests/stress/set-iterators-next-2.js: Added.
(assert):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/MapIteratorPrototype.js:
(next):
(linkTimeConstant.mapIteratorNext): Deleted.
* Source/JavaScriptCore/builtins/MapPrototype.js:
(forEach):
* Source/JavaScriptCore/builtins/SetIteratorPrototype.js:
(next):
(linkTimeConstant.setIteratorNext): Deleted.
* Source/JavaScriptCore/builtins/SetPrototype.js:
(forEach):
(isSubsetOf):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
(JSC::BytecodeIntrinsicRegistry::orderedHashTableSentinelValue):
(JSC::BytecodeIntrinsicRegistry::sentinelMapBucketValue): Deleted.
(JSC::BytecodeIntrinsicRegistry::sentinelSetBucketValue): Deleted.
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecode/SpeculatedType.cpp:
(JSC::speculationFromClassInfoInheritance):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::mapIteratorInternalFieldIndex):
(JSC::setIteratorInternalFieldIndex):
* Source/JavaScriptCore/dfg/DFGAbstractHeap.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasHeapPrediction):
(JSC::DFG::Node::hasBucketOwnerType):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::SafeToExecuteEdge::operator()):
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetMapIndexImpl):
(JSC::DFG::SpeculativeJIT::compileMapKeyIndex):
(JSC::DFG::SpeculativeJIT::compileMapValue):
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGUseKind.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGUseKind.h:
(JSC::DFG::typeFilterFor):
(JSC::DFG::isCell):
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h:
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
(Inspector::cloneMapIteratorObject):
(Inspector::cloneSetIteratorObject):
(Inspector::JSInjectedScriptHost::iteratorEntries):
* Source/JavaScriptCore/runtime/HashMapHelper.h: Added.
(JSC::areKeysEqual):
(JSC::normalizeMapKey):
(JSC::wangsInt64Hash):
(JSC::jsMapHash):
(JSC::jsMapHashImpl):
(JSC::jsMapHashForAlreadyHashedValue):
(JSC::concurrentJSMapHash):
(JSC::shouldShrink):
(JSC::shouldRehash):
(JSC::nextCapacity):
* Source/JavaScriptCore/runtime/HashMapImpl.cpp: Removed.
* Source/JavaScriptCore/runtime/HashMapImpl.h: Removed.
* Source/JavaScriptCore/runtime/HashMapImplInlines.h: Removed.
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSCJSValue.h:
(JSC::OrderedHashTableTraits::set):
(JSC::OrderedHashTableTraits::increment):
(JSC::OrderedHashTableTraits::decrement):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::createTuple):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.h:
* Source/JavaScriptCore/runtime/JSMap.h:
* Source/JavaScriptCore/runtime/JSMapInlines.h:
* Source/JavaScriptCore/runtime/JSMapIterator.cpp:
(JSC::JSMapIterator::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSMapIterator::createPair): Deleted.
* Source/JavaScriptCore/runtime/JSMapIterator.h:
* Source/JavaScriptCore/runtime/JSSet.h:
* Source/JavaScriptCore/runtime/JSSetInlines.h:
* Source/JavaScriptCore/runtime/JSSetIterator.cpp:
(JSC::JSSetIterator::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSSetIterator::createPair): Deleted.
* Source/JavaScriptCore/runtime/JSSetIterator.h:
* Source/JavaScriptCore/runtime/MapConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/MapConstructor.h:
* Source/JavaScriptCore/runtime/MapPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::createMapIteratorObject):
* Source/JavaScriptCore/runtime/OrderedHashTable.cpp: Copied from Source/JavaScriptCore/runtime/JSMapInlines.h.
(JSC::OrderedHashTable<Traits>::visitChildrenImpl):
* Source/JavaScriptCore/runtime/OrderedHashTable.h: Added.
(JSC::OrderedHashTable::OrderedHashTable):
(JSC::OrderedHashTable::offsetOfButterfly):
(JSC::OrderedHashTable::finishCreation):
(JSC::OrderedHashTable::getKeyIndex):
(JSC::OrderedHashTable::has):
(JSC::OrderedHashTable::add):
(JSC::OrderedHashTable::addNormalized):
(JSC::OrderedHashTable::remove):
(JSC::OrderedHashTable::removeNormalized):
(JSC::OrderedHashTable::size):
(JSC::OrderedHashTable::clear):
(JSC::OrderedHashTable::materializeIfNeeded):
(JSC::OrderedHashTable::storageOrSentinel):
(JSC::OrderedHashTable::storageRef):
(JSC::OrderedHashMap::OrderedHashMap):
(JSC::OrderedHashMap::getImpl):
(JSC::OrderedHashMap::get):
(JSC::OrderedHashMap::createSentinel):
(JSC::OrderedHashMap::createDeletedValue):
(JSC::OrderedHashSet::OrderedHashSet):
* Source/JavaScriptCore/runtime/OrderedHashTableHelper.h: Added.
(JSC::OrderedHashTableHelper::toNumber):
(JSC::OrderedHashTableHelper::asNumber):
(JSC::OrderedHashTableHelper::toJSValue):
(JSC::OrderedHashTableHelper::isDeleted):
(JSC::OrderedHashTableHelper::isValidTableIndex):
(JSC::OrderedHashTableHelper::invalidTableIndex):
(JSC::OrderedHashTableHelper::get):
(JSC::OrderedHashTableHelper::slot):
(JSC::OrderedHashTableHelper::set):
(JSC::OrderedHashTableHelper::setWithWriteBarrier):
(JSC::OrderedHashTableHelper::aliveEntryCountIndex):
(JSC::OrderedHashTableHelper::deletedEntryCountIndex):
(JSC::OrderedHashTableHelper::capacityIndex):
(JSC::OrderedHashTableHelper::iterationEntryIndex):
(JSC::OrderedHashTableHelper::aliveEntryCount):
(JSC::OrderedHashTableHelper::deletedEntryCount):
(JSC::OrderedHashTableHelper::usedCapacity):
(JSC::OrderedHashTableHelper::capacity):
(JSC::OrderedHashTableHelper::iterationEntry):
(JSC::OrderedHashTableHelper::incrementAliveEntryCount):
(JSC::OrderedHashTableHelper::decrementAliveEntryCount):
(JSC::OrderedHashTableHelper::incrementDeletedEntryCount):
(JSC::OrderedHashTableHelper::bucketCount):
(JSC::OrderedHashTableHelper::hashTableStartIndex):
(JSC::OrderedHashTableHelper::hashTableEndIndex):
(JSC::OrderedHashTableHelper::bucketIndex):
(JSC::OrderedHashTableHelper::dataTableSize):
(JSC::OrderedHashTableHelper::dataTableStartIndex):
(JSC::OrderedHashTableHelper::dataTableEndIndex):
(JSC::OrderedHashTableHelper::entryDataStartIndex):
(JSC::OrderedHashTableHelper::setKeyOrValueData):
(JSC::OrderedHashTableHelper::deleteData):
(JSC::OrderedHashTableHelper::addToChain):
(JSC::OrderedHashTableHelper::tableSize):
(JSC::OrderedHashTableHelper::tableSizeSlow):
(JSC::OrderedHashTableHelper::isObsolete):
(JSC::OrderedHashTableHelper::nextTable):
(JSC::OrderedHashTableHelper::setNextTable):
(JSC::OrderedHashTableHelper::deletedEntriesStartIndex):
(JSC::OrderedHashTableHelper::isCleared):
(JSC::OrderedHashTableHelper::setClearedTableSentinel):
(JSC::OrderedHashTableHelper::tryCreate):
(JSC::OrderedHashTableHelper::copyImpl):
(JSC::OrderedHashTableHelper::copy):
(JSC::OrderedHashTableHelper::rehash):
(JSC::OrderedHashTableHelper::clear):
(JSC::OrderedHashTableHelper::find):
(JSC::OrderedHashTableHelper::expandIfNeeded):
(JSC::OrderedHashTableHelper::addImpl):
(JSC::OrderedHashTableHelper::add):
(JSC::OrderedHashTableHelper::addNormalized):
(JSC::OrderedHashTableHelper::shrinkIfNeed):
(JSC::OrderedHashTableHelper::removeImpl):
(JSC::OrderedHashTableHelper::remove):
(JSC::OrderedHashTableHelper::removeNormalized):
(JSC::OrderedHashTableHelper::transit):
(JSC::OrderedHashTableHelper::transitAndNext):
(JSC::OrderedHashTableHelper::getKey):
(JSC::OrderedHashTableHelper::getValue):
(JSC::OrderedHashTableHelper::nextAndUpdateIterationEntry):
(JSC::OrderedHashTableHelper::getIterationEntry):
(JSC::OrderedHashTableHelper::getIterationEntryKey):
(JSC::OrderedHashTableHelper::getIterationEntryValue):
* Source/JavaScriptCore/runtime/SetConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/SetConstructor.h:
* Source/JavaScriptCore/runtime/SetPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::createSetIteratorObject):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::orderedHashTableDeletedValueSlow):
(JSC::VM::orderedHashTableSentinelSlow):
(JSC::VM::visitAggregateImpl):
(JSC::VM::sentinelSetBucketSlow): Deleted.
(JSC::VM::sentinelMapBucketSlow): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::orderedHashTableDeletedValue):
(JSC::VM::orderedHashTableSentinel):
(JSC::VM::sentinelSetBucket): Deleted.
(JSC::VM::sentinelMapBucket): Deleted.
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
* Source/JavaScriptCore/runtime/WeakMapImplInlines.h:
* Source/JavaScriptCore/runtime/WeakSetPrototype.cpp:
* Source/WebCore/bindings/js/JSDOMMapLike.cpp:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::serialize):

Canonical link: https://commits.webkit.org/280734@main
